authorize/log.go

```diff
@@ -92,56 +73,140 @@ func (a *Authorize) logAuthorizeCheck(
 	if err != nil || len(res.GetRecords()) == 0 {
-		return evt
+		return nil
 	}
 
 	impersonatedUserMsg, err := res.GetRecords()[0].GetData().UnmarshalNew()
 	if err != nil {
-		return evt
+		return nil
 	}
 	impersonatedUser, ok := impersonatedUserMsg.(*user.User)
 	if !ok {
-		return evt
+		return nil
 	}
-	evt = evt.Str("impersonate-email", impersonatedUser.GetEmail())
+	email := impersonatedUser.GetEmail()
 
-	return evt
+	return &impersonateDetails{
+		sessionID: sessionID,
+		userID:    userID,
+		email:     email,
+	}
+}
+
+func populateLogEvent(
+	ctx context.Context,
+	field log.AuthorizeLogField,
+	evt *zerolog.Event,
+	in *envoy_service_auth_v3.CheckRequest,
+	s sessionOrServiceAccount,
+	u *user.User,
+	hdrs map[string]string,
+	impersonateDetails *impersonateDetails,
+) *zerolog.Event {
+	if debugField, ok := field.IsForDebugOnly(); ok {
+		if zerolog.GlobalLevel() > zerolog.DebugLevel {
+			// ignore because we're not logging debug messages
+			return evt
+		}
+		field = debugField
+	}
+
+	switch field {
+	case log.AuthorizeLogFieldCheckRequestID:
+		return evt.Str(string(field), hdrs["X-Request-Id"])
+	case log.AuthorizeLogFieldEmail:
+		return evt.Str(string(field), u.GetEmail())
+	case log.AuthorizeLogFieldHeaders:
+		return evt.Interface(string(field), hdrs)
+	case log.AuthorizeLogFieldHost:
+		return evt.Str(string(field), in.GetAttributes().GetRequest().GetHttp().GetHost())
+	case log.AuthorizeLogFieldImpersonateEmail:
+		if impersonateDetails != nil {
+			evt = evt.Str(string(field), impersonateDetails.email)
+		}
+		return evt
+	case log.AuthorizeLogFieldImpersonateSessionID:
+		if impersonateDetails != nil {
+			evt = evt.Str(string(field), impersonateDetails.sessionID)
+		}
+		return evt
+	case log.AuthorizeLogFieldImpersonateUserID:
+		if impersonateDetails != nil {
+			evt = evt.Str(string(field), impersonateDetails.userID)
+		}
+		return evt
+	case log.AuthorizeLogFieldIP:
+		return evt.Str(string(field), in.GetAttributes().GetSource().GetAddress().GetSocketAddress().GetAddress())
+	case log.AuthorizeLogFieldMethod:
+		return evt.Str(string(field), in.GetAttributes().GetRequest().GetHttp().GetMethod())
+	case log.AuthorizeLogFieldPath:
+		return evt.Str(string(field), stripQueryString(in.GetAttributes().GetRequest().GetHttp().GetPath()))
+	case log.AuthorizeLogFieldQuery:
+		return evt.Str(string(field), in.GetAttributes().GetRequest().GetHttp().GetQuery())
+	case log.AuthorizeLogFieldRequestID:
+		return evt.Str(string(field), requestid.FromContext(ctx))
+	case log.AuthorizeLogFieldServiceAccountID:
+		if sa, ok := s.(*user.ServiceAccount); ok {
+			evt = evt.Str(string(field), sa.GetId())
+		}
+		return evt
+	case log.AuthorizeLogFieldSessionID:
+		if s, ok := s.(*session.Session); ok {
+			evt = evt.Str(string(field), s.GetId())
+		}
+		return evt
+	case log.AuthorizeLogFieldUser:
+		return evt.Str(string(field), u.GetId())
+	default:
+		return evt.Str(string(field), "<<UNKNOWN AUTHORIZE LOG FIELD>>")
```

Comment:
This feels a little awkward to me. Should we add logic to Options.Validate() to verify that there aren't any unknown log fields? Would it make sense to silently skip unknown fields here if we logged a warning for any unknown fields from Options.Validate()?

Reply:
Yes I can add code to validate the options, in which case silently ignoring is ok.